Classification app

erikwaskiewicz opened the pull request:
Initial PR for the classification app, not all diff will appear here but dont want to merge to dev yet

- all classifications page
  - shows all classifications, formatting is basic at the mo, will eventually probably hide all completed classifications and make a separate page to show those. Mimics the equivalent page in SVD with whos assigned etc. Button to add new classification is going to be removed, this is just to make dev quicker

- classify page
  - variant info tab - just a click button to complete, may be fields edited in future (such as specific tumour type). Reopen button resets everything else on the check
  - previous classifications tab - quite a complex one, displays previous classifications in the specific tumour type (not actually doing that yet, there's a TODO that it needs to be specific to tumour type in the code, logic will be the same though), and displays all previous classifications in a modal. If previous is there and selected then it saved to the model as a foreign key and doesnt make code obejcts, vice versa if full classification done. 2nd checker has to do the same otherwise they need to send check back
  - classification tab - should add up correctly, only issue at the mo is that the override dropdown has every single option available (e.g. oncogenic for ACMG), needs to be moved from settings to models. Some GUI things left to do but they're minor - comments, guidence modal, annotation HTML blocks
  - finalise tab - shows all classifications and choose the next step, error handling at the mo is limited to an error if you select the wrong thing (e.g. will error if you try to finalise with only one check), this could be made nicer but works for now


There's still a little bit of SVIG specific code left but its fairly minimal now, it should be a fairly straightforward PR to fix those seperately. Thats mostly a few specific model functions, I'm happy that the views and forms are good

Any issues let me know and we can talk it through :)

Istaisa commented on env.yaml:
Environment builds from scratch on dev machine.

Istaisa commented on classify/fixtures/setup_acgs_fixtures.json:
this needs deleting - ACGS and SVIG are both in the other setup fixtures

Istaisa replied:
done in [ac6c40d](https://github.com/AWGL/somatic_db/pull/164/commits/ac6c40d21178bf085fce38c862b4822d96a30833)

Istaisa commented on classify/fixtures/setup_fixtures.json:
Should we add an issue to add in the CanVig guidelines?

Istaisa commented on classify/views.py:
```diff
@@ -0,0 +1,225 @@
+        if "complete_classification" in request.POST:
+            complete_classification_form = CompleteClassificationForm(request.POST)
+            if complete_classification_form.is_valid():
+                override = complete_classification_form.cleaned_data["override"]
```
If you override the classification then unlock it and use the original, it still flags the classification as overriden

Istaisa replied:
done in [b5a70b9](https://github.com/AWGL/somatic_db/pull/164/commits/b5a70b93af9a8e125df6b82450aaa0baaa7c7432)

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,846 @@
+    Codes that can be applied
+    """
+    code = models.CharField(max_length=10, unique=True)
+    pathogenic_or_benign = models.CharField(max_length=1)
```
don't know if we want to rename this field because it's also or oncogenic, and also if we should make this a choicefield to restrict to B/P/O

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,846 @@
+
+    @transaction.atomic
+    def reopen_classification_tab(self):
+        """reset the classifications tab"""
```
I think the is there a manual override bit probably needs resetting here

Istaisa commented on classify/templatetags/custom_tags.py:
```diff
@@ -0,0 +1,78 @@
+@stringfilter
+def colour_by_guideline(value):
+    """
+    colour different guidelines differently, e.g. ACMG/ SVIG
```
do we want to make this a more general germline/somatic (this might be a field that needs adding in to the guideline model) so that any future and/or gene specific things go the right colour

Istaisa replied:
done in [b42ea1b](https://github.com/AWGL/somatic_db/pull/164/commits/b42ea1b0ff97131be8bac01944897a76ff9b2eb0)

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+    Strengths at which the criteria can be applied at
+    """
+    strength = models.CharField(max_length=20)
+    shorthand = models.CharField(max_length=2)
```
this applies to everywhere we've got a shorthand - not sure we still need it now everything's in models but also i don't want to mess with the ajax

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+        return self.strength.title().replace("_", " ")
+
+
+class ClassifyVariant(models.Model):
```
should probably have a unique_together on this

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+            else:
+                return None, False
+
+            # TODO maybe a warning if multiple pending classifications for same variant?
```
this sounds like a good idea

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+            return True, None
+
+    @transaction.atomic
+    def signoff_check(self, current_check, next_step):
```
If you're on the first check and choose the complete/send back options it errors as expected but also  closes off the check

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+    """
+    Variants added manually directly to classify
+    """
+    pass
```
do we want info like tumour type in here?

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+
+        # work out class from score counter
+        class_dict = self.classification.guideline.create_final_classification_ordered_dict()
+        # loop through in order until the score no longer meets the threshold
```
I can't see it but if for whatever reason we use this with guidelines where 'Benign' is not the most benign option this'll need a tweak

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+        try:
+            return self.final_class.final_classification
+        except AttributeError:
+            return "Pending"
```
I had to add this to make things work after I moved things from settings to models - feel free to change the display text it throws this error on new unclassified checks

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+        else:
+            return f"{self.code.code}_NA"
+
+    def get_code_type(self):
```
I think we do this bit of code over a few places - could generalise by making it a static method

Istaisa commented on classify/models.py:
```diff
@@ -0,0 +1,871 @@
+            return "Not applied"
+
+
+'''class AbstractComment(models.Model):
```
remove these?

Istaisa reviewed (changes requested):
Initial code review for classification app. Overall looking really good, can follow the logic for everything and user testing so far works as intended.

Not reviewing known outstanding features:

- Linking in the variants from the other apps to classifications, removing the hardcoded variant 
- Adding in tumour type
- Tests
- Known clearup and refactoring in outstanding TODOs
- Documentation

Environment builds from scratch on dev laptop.

Tested different combinations of codes and the score totals and final classification seems to be correct for both SVIG and ACGS criteria. Functionality around sending for an extra check / finalising with 2 checks / sending a check back all appear to work correctly.

Main comments for this PR:

- I've moved the final classification logic from settings to models and done a couple of other minor tweaks - please ask if anything doesn't make sense. I don't think I broke the colours this time but also I might have. Change to the models and the fixtures - have generated new ones
- There's a tiny display issue on the classify page when you're doing the second check where the previous classifier's rating doesn't show correctly as the code is repeated (e.g. O5 O5 (+2) instead of O5 moderate {+2))
- Checks are closing off incorrectly if a first check is sent back or sent to close
- 2nd check can be finalised if checker 1 and checker 2 disagree

erikwaskiewicz commented:
Reviewed all changes from last review and all looks good, just a couple of small bug fixes where the get_final_class_display function was being used instead of the model itself

Also fixes other issues:
- added tumour type and linked to previous classifications page, currently directly to classify instance model but let me know if you have thoughts on where this would be better situated, remaining logic should stay the same regardless of where it sits though
- added a training flag and user groups for signed off users, mostly copied fro your other branch but lists in Guideline model to handle being signed off for e.g. SVIG but not ACMG
-  Bugs around closing analysis fixed and now checks to make sure the las checks agree and are different people (commented out to make testing easier). Disabled being able to reopen a case if its been used as a previous classificaion
- Added a linked classifications box but just a placeholder for now, just shows a queryset for each guideline containing all classifications on the exact same variant instance. Will discuss how to format this with Sian L 
- Updated fixtures

These are outstanding but if it looks good then I'm happy that this branch can be merged and we pick them up in a new branch
- Comments
- Unit tests
- Finish linked classifications bit from above

(also just seen this is merging into main, dont want to change the base now incase it confuses things but will do when changes are approved)

Istaisa reviewed (approved):
Existing bugs fixed and new changes work as expected - good to merge for initial testing.
I'll work on the unit tests and let you review those.
Would be good if we added some tumour types in to the fixtures, but assume we're going to pull this from analysis?

Istaisa commented:
Most of the unit tests finished, some outstanding ones from the classify instance. Will see if I can finish them between meetings Monday/Tuesday.

Istaisa commented:
unit tests are mostly done and mostly work

Istaisa reviewed (approved):
Updated unit tests logical and pass.
Raised issues for outstanding work #202 #203 
Added a test to cover the views. Good to merge
